chainer/testing/function_link.py

```diff
@@ -0,0 +1,1116 @@
+              [
+                  {},  # CPU
+                  {'use_cuda': True},  # GPU
+              ])
+            class TestLinear(chainer.testing.LinkTestCase):
+
+                param_names = ['W', 'b']
+
+                def generate_params(self):
+                    initialW = numpy.random.uniform(
+                        -1, 1, (3, 2)).astype(numpy.float32)
+                    initial_bias = numpy.random.uniform(
+                        -1, 1, (3,)).astype(numpy.float32)
+                    return initialW, initial_bias
+
+                def generate_inputs(self):
+                    x = numpy.random.uniform(
+                        -1, 1, (1, 2)).astype(numpy.float32)
+                    return x,
+
+                def create_link(self, initializers):
+                    initialW, initial_bias = initializers
+                    link = chainer.links.Linear(
+                        2, 3, initialW=initialW, initial_bias=initial_bias)
+                    return link
+
+                def forward(self, link, inputs, device):
+                    x, = inputs
+                    return link(x),
+
+                def forward_expected(self, inputs, params):
+                    x, = inputs
+                    W, b = params
+                    expected = x.dot(W.T) + b
+                    return expected,
+
+    .. seealso::
+        :class:`~chainer.testing.LinkInitializersTestCase`
+        :class:`~chainer.testing.FunctionTestCase`
+
+    """
+
+    check_forward_options = {}
+    check_backward_options = {}
+    skip_forward_test = False
+    skip_backward_test = False
+    dodge_nondifferentiable = False
+
+    def forward_expected(self, inputs, params):
+        raise NotImplementedError('forward_expected() is not implemented.')
+
+    def generate_grad_outputs(self, outputs_template):
+        grad_outputs = tuple([
+            numpy.random.uniform(-1, 1, a.shape).astype(a.dtype)
+            for a in outputs_template])
+        return grad_outputs
+
+    def test_forward(self, backend_config):
+        """Tests forward computation."""
+
+        if self.skip_forward_test:
+            raise unittest.SkipTest('skip_forward_test is set')
+
+        self.before_test('test_forward')
+
+        inits = self._generate_params()
+        link = self._create_link(inits, backend_config)
+
+        inputs_np = self._generate_inputs()
+        inputs_xp = backend_config.get_array(inputs_np)
+        inputs_xp = self._to_noncontiguous_as_needed(inputs_xp)
+        input_vars = tuple([chainer.Variable(i) for i in inputs_xp])
+        # Compute forward of the link and initialize its parameters.
+        output_vars = self._forward(link, input_vars, backend_config)
+        outputs_xp = [v.array for v in output_vars]
+
+        # Expected outputs are computed on the CPU so inputs and parameters
+        # must be transferred.
+        cpu_device = backend.CpuDevice()
+        params = _get_link_params(link, self.param_names)
```

Comment:
I think all the parameter in the link is needed (not only those specified in `self.param_names`).

Reply:
Agreed offline to pass the link object (transferred to CPU).

Reply:
Fixed.